src/common/mfu_daos.c

```diff
@@ -1244,6 +1309,90 @@ int daos_parse_epc_str(
     return 0;
 }
 
+#ifdef HDF5_SUPPORT
+static int serialize_daos_metadata(char *filename,
+                                   daos_args_t* da)
+{
+    int    rc = 0;
+    hid_t  status = 0;
+    struct hdf5_args hdf5;
+
+    rc = asprintf(&filename, "%s%s%s", da->daos_preserve_path, "/", "daos_metadata.hdf5");
+    printf("Writing metadata file: %s\n", filename);
+    if (rc == -1) {
+        MFU_LOG(MFU_LOG_ERR, "failed to write filename");
+        goto out;
+    }
+
+    /* TODO: much of this HDF5 setup should get moved into the HDF5 user
+     * interface */
+    hdf5.file = H5Fcreate(filename, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+    if (hdf5.file < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create hdf5 file");
+        goto out;
+    }
+    rc = cont_serialize_props(&hdf5, da->src_coh);
+    if (rc != 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to serialize cont layout: "DF_RC, DP_RC(rc));
+        goto out;
+    }
+
+    /* create User Attributes Dataset in daos_metadata file */
+    hdf5.usr_attr_memtype = H5Tcreate(H5T_COMPOUND, sizeof(usr_attr_t));
+    if (hdf5.usr_attr_memtype < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create user attr memtype");
+        goto out;
+    }
+    hdf5.usr_attr_name_vtype = H5Tcopy(H5T_C_S1);
+    if (hdf5.usr_attr_name_vtype < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create user attr name vtype");
+        goto out;
+    }
+    status = H5Tset_size(hdf5.usr_attr_name_vtype, H5T_VARIABLE);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create user attr name vtype");
+        goto out;
+    }
+    hdf5.usr_attr_val_vtype = H5Tvlen_create(H5T_NATIVE_OPAQUE);
+    if (hdf5.usr_attr_val_vtype < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to create user attr val vtype");
+        goto out;
+    }
+    status = H5Tinsert(hdf5.usr_attr_memtype, "Attribute Name",
+                       HOFFSET(usr_attr_t, attr_name), hdf5.usr_attr_name_vtype);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to insert user attr name");
+        goto out;
+    }
+    status = H5Tinsert(hdf5.usr_attr_memtype, "Attribute Value",
+                       HOFFSET(usr_attr_t, attr_val), hdf5.usr_attr_val_vtype);
+    if (status < 0) {
+        MFU_LOG(MFU_LOG_ERR, "failed to insert user attr val");
+        goto out;
+    }
```

Comment:
@dsikich I think we should set rc = 1 for these failed hdf5 calls so the caller properly knows it failed.